### DISCUSSION TITLE
Device topologies and adapter operation modes

### DISCUSSION THREAD

**Discussion by MyHomeMyData** [MyHomeMyData/ioBroker.e3oncan#34] [Show and tell]
Adapter e3oncan offers two basically different modes of operation which may be used in parallel.

To decide for best mode(s) it's imporant to know about the device topology of own setup. The most important criterion is the networking of the devices with each other via CAN. Here's an example for two different setups with energy meter, Vitocharge and Vitocal:

![E3_devices_not_connected](https://github.com/MyHomeMyData/ioBroker.e3oncan/assets/144950531/1141ee76-31f0-4e23-87a3-d16873f3b5ee)

![E3_devices_connected](https://github.com/MyHomeMyData/ioBroker.e3oncan/assets/144950531/4a89fc4d-56a9-4465-b2c0-ef4025294fec)

In second setup Vitocharge and Vitocal are connected via external CAN bus. Vitocal operated as master, Vitocharge in slave mode. This configuration can be created by the heating engineer or Viessmann support during commissioning of the devices. Simply connecting the devices via CAN is not effective and would result in error messages.

Dependent on your setup following operation modes may be used with e3oncan adapter:

![image](https://github.com/user-attachments/assets/19b503ff-cc8c-4680-9c2a-e23a034d8bd2)

To apply external and internal CAN in parallel you have to use two USB2CAN adapters.

**Comment by Dreman65** [comment]
Meine derzeitige Konfiguration: Gastherme Vitodens 300-W-B3HG. Can-Adapter (1-Kanal CanHat) am **internen** Can-Anschluss der Vitodens (an Stecker X5 über Adapter). Damit funktioniert sowohl Collect als auch Lesen und Schreiben über UDSonCAN (einschließlich Schreiben über Service77) zuverlässig. Im Zuge der aktuellen Umstellung des Raspi (von 3B, 32 bit Buster auf 4B, 64 bit Bookworm, jeweils ohne Desktop) hab ich mir gleich einen 2-Channel CanHat besorgt, mit dem Hintergedanken, den **externen** CAN (Stecker 91) für UDSonCan und den **internen** CAN (Stecker X5) für Collect jeweils getrennt zu verwenden (can0 und can1). Jetzt kommen mir jedoch Zweifel an meiner Idee, da dann ja wahrscheinlich das automatisierte Schreiben mittels Service77 nicht mehr funktioniert? Liege ich damit richtig? Oder gibt einen workaround?

**Reply by MyHomeMyData** [reply]
Freut mich, dass alles gut funktioniert!

Ja, da liegst Du richtig. Service 77 läuft auf dem externen Bus ins Leere. Auf entsprechende Anfragen kommt auf dem externen Bus schlicht keine Antwort. Es gibt aber doch gar keinen Grund, Collect und UDSonCAN zu trennen, oder? Collect ist 100% passiv, sendet also kein einziges Byte auf den Bus. Es kann sich also mit UDSonCAN nicht ins Gehege kommen.

**Reply by Dreman65** [reply]
Danke für die schnelle Antwort. Grund gibt es eigentlich keinen, war halt so eine Idee, da mir jetzt 2 CANs zur Verfügung stehen :-).

**Comment by cbh67** [comment]
Ich nutze die sehr gute Lösung jetzt einige Monate mit 2 Adaptern jeweils am externen CAN-Bus an Vitodens 200W und Vitocal 250A, da es Viessmann bisher nicht schafft die beiden Geräte per CAN-Bus miteinander kommunizieren zu lassen. Eine Alternative wurde in Aussicht gestellt, ist aber noch nicht realisiert.
Vielen Dank an den/die Entwickler für das Engagenment offensichtliche Herstellerdefizite auszugleichen.
Leider funktionieren seit letztem Wochenende einige Datenpunkte nicht mehr und werden bei Abfrage mit 0x22 bzw. 0x31 Fehlern angezeigt. Eine Liste kann ich liefern. 
Ich habe die zunächst aus der Abfrage entfernt, um die regelmäßigen Fehlermeldungen zu unterbinden.
Bei der Vitocal handelt es sich meist um Datenpunkte zum Kältekreis, die scheinbar am externen CAN-Bus nicht mehr angeboten werden. Leider konnte ich keine Hinweise auf ein evtl. Firmwareupdate der Anlagen finden. Gibt es ähnliche Erkenntnisse bei anderen Nutzern?

**Reply by MyHomeMyData** [reply]
@cbh67 Mir sind keine ähnliche Probleme bekannt. Meine Vitocal 252A funktioniert weiterhin ohne Fehlermeldungen. Bitte schick mal ein paar Beispiele.

Hast Du schon das AEG-Verfahren probiert: Ausschalten-Einschalten-Gut?

**Comment by cbh67** [comment]
Vielen Dank für die schnelle Antwort.
Ja habe ich schon versucht. Adapter läuft auf raspi, alles auf aktuellem Stand und neu gebootet. Auch iobroker auf Synology neues Dockerimage erstellt. Vitodens ausgeschaltet und wieder neu gestartet. Keine Änderung.
DP 331,364,522 2353 und 544,545,929,1342,1343,1344,1345,1346
Vitocal sind mehr, eigentlich die Visualisierung Kältekreislauf fast komplett. Beispielsweise die Fans 1775,1776.
Rest kommt morgen.
VG Bernd 

Am 19. April 2025 17:23:58 MESZ schrieb MyHomeMyData ***@***.***>:

**Reply by Dreman65** [reply]
@cbh67  Hab zwar eine 300-W-B3HG und frage auf dem internen CAN ab, aber alle die von dir oben angegebenen DP für die Vitodens laufen bei mir nach wie vor fehlerfrei (Abfragezyklus 60 Sekunden) und ohne Fehlermeldungen (e3oncan 0.10.8)

**Reply by MyHomeMyData** [reply]
@cbh67 Auf meiner Vitocal laufen alle genannten DIDs fehlerfrei durch.
Eine mögliche Ursache könnte sein, dass ein anderer Prozess parallel auf dem Gerät Abfragen macht. Z,B, ein parallel laufendes open3e oder eine weitere e3oncan-Instanz.

Anderer Denkansatz: Als das seltsame Verhalten begonnen hat, was hast Du da vorher gemacht/geändert?

Was steht im State `e3oncan.0.HPMUMASTER_0x680.info.statUDS`? Bei mir sieht das so aus:
```
{
  "state": "active",
  "CANdevAddr": "0x680",
  "cntCommTotal": 931300,
  "cntCommOk": 931286,
  "cntCommNR": 14,
  "cntCommTimeout": 0,
  "cntCommBadProtocol": 0,
  "cntCommFailedPerDid": {},
  "cntTooBusy": 0,
  "replyTime": {
    "min": 2,
    "max": 203,
    "mean": 8
  },
  "nextTs": 1745091972707,
  "tsMinStep": 5000
}
```

**Reply by MyHomeMyData** [reply]
@cbh67 Wenn das Problem weiter besteht, bitte einen CAN-Dump erstellen und die Log-Datei hier posten:

* can utils installieren: `sudo apt install can-utils`
* can dump starten: `candump -tA can0 | grep "  6[89]0  " > candump_nr.log`. can0 ggf. anpassen.
* Warten, bis fehlerhafte Übertragungen stattgefunden haben. Dann candump mit <STRG>-C abbrechen.

**Comment by cbh67** [comment]
Problem scheint gefunden zu sein. Nachdem ich alle Logs durchsucht und mir nochmal den Ablauf letzte Woche angesehen hatte, kam ich zu folgendem Schluss:
Raspi reagierte nicht mehr und musste vom Strom getrennt werden. Danach sind die CAN-Bus-Kanäle offenbar vertauscht worden. Vermutlich hatte ich bei der mehrere Wochen zeitversetzten Inbetriebnahme der beiden USB2CAN-Adapter den Raspi nicht nochmal gebootet. Jetzt nach dem Kaltstart wurden offenbar die Adapter an den USB-Ports anders zugeordnet. Damit wurde Vitocal mit den Parametern von Vitodens abgefragt und umgekehrt. Wo die Datenpunkte gleich waren, gab es kein offensichtliches Problem. Nur bei denen, die es jeweils nur auf einer Seite gab, kam eine Fehlermeldung. Inzwischen war in den Datenpunkten eine Mischung aus beiden entstanden, so dass eine Neuinstallation beider e3oncan-Adapter die sinnvollste Lösung war. Bei gleicher Reihenfolge der Installation konnten auch die Objekte in der Visualisierung ohne Änderung weiter genutzt werden.
Vielen Dank für die Hinweise und die gute Arbeit.
Falls mal Testbedarf sein sollte, gern melden. 
Ich muss die Konfiguration voraussichtlich im Mai/Juni überarbeiten,  wenn die beiden Wärmequellen dann nach einem Jahr endlich gekoppelt werden sollen/können. Voraussichtlich aber eben über den PlusBus.

Am 20. April 2025 08:49:36 MESZ schrieb MyHomeMyData ***@***.***>:

**Reply by MyHomeMyData** [reply]
@cbh67 Danke für die Rückmeldung. Damit ist das Verhalten nachvollziehbar.

Die Zuordnung der USB-Adapter geschieht bei Start zufällig. Deshalb solltest Du die Zuordnung festzurren. Eine Möglichkeit das zu tun, ist [hier](https://github.com/open3e/open3e/blob/master/bestPractices/USB%20Adapter%20and%20udev.md) beschrieben. Mit dieser Methode werden die Adapter auch gleich korrekt initialisiert. Sowohl beim Start, als auch beim Ab- und Anstecken im laufenden Betrieb.
